pkg/controller/nodes/dynamic/handler.go

```diff
@@ -191,11 +191,16 @@ func (d dynamicNodeTaskNodeHandler) Abort(ctx context.Context, nCtx handler.Node
 
 // This is a weird method. We should always finalize before we set the dynamic parent node phase as complete?
 func (d dynamicNodeTaskNodeHandler) Finalize(ctx context.Context, nCtx handler.NodeExecutionContext) error {
+	// We should always finalize the parent node success of failure.
+	// If we use the state to decide the finalize then we will never invoke the finalizer for the parent.
+	logger.Infof(ctx, "Finalizing Parent node")
+	if err := d.TaskNodeHandler.Finalize(ctx, nCtx); err != nil {
+		logger.Errorf(ctx, "Failed to finalize Dynamic Nodes Parent.")
+		return err
```

Comment:
Should we still optimistically finalize the could nodes?

Reply:
i dont understand, do you mean in failure scenario. Child nodes are actually finalized everytime they complete. Abort is the scenario where a failure to finalize parent may cause child nodes to not be finalized, (persistent failure). Maybe we could change the order

Reply:
What do you mean by optimistically haytham?